src/js_types.hpp

```diff
@@ -488,9 +507,92 @@ inline typename T::Value Value<T>::from_mixed(typename T::Context ctx, const uti
         return from_string(ctx, value.get<StringData>().data());
     case type_Binary:
         return from_binary(ctx, value.get<BinaryData>());
-    default:
-        throw std::invalid_argument("Value not convertible.");
+
+    case type_Link:
+    case type_LinkList:
+    case type_OldDateTime:
+    case type_OldTable:
+    case type_OldMixed:
+        break;
+    }
+    throw std::invalid_argument("Value not convertible.");
+}
+
+template<typename T>
+inline typename T::Value Value<T>::from_bson(typename T::Context ctx, const bson::Bson& value) {
+    using Type = bson::Bson::Type;
+
+    switch (value.type()) {
+    case Type::MinKey:
+        return Object<T>::create_bson_type(ctx, "MinKey", {});
+    case Type::MaxKey:
+        return Object<T>::create_bson_type(ctx, "MaxKey", {});
+    case Type::Null:
+        return from_null(ctx);
+    case Type::Bool:
+        return from_boolean(ctx, value.operator bool());
+    case Type::Double:
+        return from_number(ctx, value.operator double());
+    case Type::Int32:
+        // All int32 values can be precisely represented as a double
+        return from_number(ctx, double(value.operator int32_t()));
+    case Type::Int64:
+        // int64 needs special handling.
+        // XXX Consider using js numbers for values within [-2^52, 2^52)
+        return Object<T>::create_bson_type(ctx, "Long", {
+            Value<T>::from_string(ctx, std::to_string(value.operator int64_t())),
+        });
+    case Type::Decimal128:
+        return from_decimal128(ctx, value.operator Decimal128());
+    case Type::ObjectId:
+        return from_object_id(ctx, value.operator ObjectId());
+    case Type::Timestamp:
+        // XXX this is wrong. Should not be creating a JS Date!
+        // See https://github.com/realm/realm-object-store/pull/1005
+        return from_timestamp(ctx, value.operator Timestamp());
+    case Type::Datetime:
+        // TODO adjust after https://github.com/realm/realm-object-store/pull/1005
+        return from_timestamp(ctx, Timestamp(value.operator bson::Datetime().seconds_since_epoch, 0));
+    case Type::String:
+        return from_string(ctx, value.operator const std::string&());
+    case Type::Binary: {
+        // XXX Should this be realm binary or bson binary?
```

Comment:
We should align with the other SDKs.

Reply:
Seems like Cocoa is using native and Java is using BSON. I'm going with the BSON since that gives us a place to put the binary subtype once it is exposed from object store. Users can still get to the js-native buffer by adding `.buffer` to `user.custom_data.my_binary_field`.

Also, if I'm reading this right, we already are inconsistent between engines. v8 uses an ArrayBuffer for binary fields, while jsc uses Uint8Array. This will use Uint8Array everywhere.